keps/sig-scheduling/5598-opportunistic-batching/README.md

```diff
@@ -0,0 +1,967 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes an opportunistic batching mechanism in the scheduler to improve performance of scheduling many compatible pods at once, and to begin building the infrastructure required for gang scheduling. To implement this mechanism we propose the following additions:
+
+ - **Pod scheduling signature:** A signature that captures the properties of a pod that impact scoring and feasibility.
+ - **Batching mechanism:** A mechanism to reuse the scheduling output from one pod to set the nominated node name for multiple subsequent pods with matching scheduling signatures.
+ - **Opportunistic batching:** Transparent inclusion of the batching mechanism in the scheduler to improve the performance of targeted workloads that could benefit from it.
+
+## Motivation
+
+Today our scheduling algorithm is O(num pods x num nodes). As the size of clusters and jobs continue to increase, this leads to low performance when scheduling or rescheduling large jobs. This increases user cost and slows down user jobs, both unpleasant impacts. Optimizations like this one have the potential to dramaticly reduce the cost of scheduling in these scenarios.
+
+We are also working on gang scheduilng (in addition to other forms of multi-pod scheduling), which will give us a way to consider multiple pods at the same time. "Opportunistic batching" provides a starting point for these mechanisms by providing signatures and batching, both necessary foundational mechanisms, and including them initially in a simple way.
+
+Another change is the shift towards 1-pod-per-node in batch and ML environments. Many of these environments (among others) only attempt to run a single user pod on each node, along with a complement of daemon set pods. This simplifies our scheduling needs significantly, as it allows to reuse not only filtering, but also scoring results.
+
+### Goals
+
+ * Improve the performance of scheduling large jobs on large clusters where the constraints are simple.
+ * Begin building infrastructure to support gang scheduling and other "multi-pod" scheduling mechanisms.
+ * Ensure that the infrastructure we build is maintainable as we update, add and remove plugins.
```

Comment:
Similar to this goal, but we also should clarify that we must keep supporting schedulers with custom plugins. like, we must not implement anything from the assumption of in-tree plugins directly onto the scheduling framework runtime.